Tribler/community/dht/routing.py

```diff
@@ -0,0 +1,192 @@
+import time
+import random
+import binascii
+
+from threading import RLock
+from socket import inet_aton
+
+import datrie
+
+from Tribler.pyipv8.ipv8.peer import Peer
+
+NODE_STATUS_GOOD = 2
+NODE_STATUS_UNKNOWN = 1
+NODE_STATUS_BAD = 0
+
+MAX_BUCKET_SIZE = 20
+
+
+def id_to_binary_string(node_id):
+    return format(int(node_id.encode('hex'), 16), '0160b')
+
+
+def distance(a, b):
+    return int(a.encode('hex'), 16) ^ int(b.encode('hex'), 16)
+
+
+def calc_node_id(ip, mid):
+    ip_bin = inet_aton(ip)
+    ip_mask = '\x03\x0f\x3f\0xff'
+    ip_masked = ''.join([chr(ord(ip_bin[i]) & ord(ip_mask[i])) for i in range(4)])
+
+    crc32_unsigned = binascii.crc32(ip_masked) % (2 ** 32)
+    crc32_bin = ('%08x' % crc32_unsigned).decode('hex')
+
+    return crc32_bin[:3] + mid[:17]
+
+
+class Node(Peer):
+
+    def __init__(self, *args, **kwargs):
+        super(Node, self).__init__(*args, **kwargs)
+        self.bucket = None
+        self.last_response = 0
+        self.last_query = 0
+        self.failed = 0
+        self.rtt = 0
+
+    @property
+    def id(self):
+        return calc_node_id(self.address[0], self.mid)
+
+    @property
+    def last_contact(self):
+        return max(self.last_response, self.last_query)
+
+    @property
+    def status(self):
+        # A good node is a node has responded to one of our queries within the last 15 minutes, or has ever responded
+        # to one of our queries and has sent us a query within the last 15 minutes. This is the same logic as
+        # used in BEP-5
+        now = time.time()
+        if ((now - self.last_response) < 15 * 60) or (self.last_response > 0 and (now - self.last_query) < 15 * 60):
+            return NODE_STATUS_GOOD
+        elif self.failed >= 2:
+            return NODE_STATUS_BAD
+        return NODE_STATUS_UNKNOWN
+
+    def distance(self, other_node):
+        return distance(self.id, other_node.id)
+
+
+class Bucket(object):
```

Comment:
Could you please add a small docstring under the most important classes to explain what it represents?

Reply:
Done.